### PR TITLE
Expose graph ModelPath to TensorRT shared library

### PR DESCRIPTION
**Pull request by stevenlix**
1. Add graphviewer ModelPath to TRT shared library
   TRT needs model path info when runs models with external weights
2. Use GenerateMetaDefId to generate unique TRT kernel name and engine name
   Remove TRT internal id generation and use ExecutionProvider's GenerateMetaDefId as a general solution

### CODE REVIEW

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 1]
```diff
@@ -152,7 +151,7 @@ class TensorrtExecutionProvider : public IExecutionProvider {
   other execution provider.
   */
   SubGraphCollection_t GetSupportedList(SubGraphCollection_t supported_nodes_list, int iterations, const int max_iterations,
-                                        const GraphViewer& graph, bool* early_termination) const;
+                                        const GraphViewer& graph, const PathString model_path, bool* early_termination) const;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h:154]: can this be passed by const ref?

Comment (stevenlix) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: Sure. Changed to pointer

Comment (pranavsharma) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: If you don't expect it to be nullptr, it's better to pass it by const-ref.

Comment (jywu-msft) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h]: change to const PathString&

File: onnxruntime/core/providers/shared_library/provider_interfaces.h [hunk 2]
```diff
@@ -1000,6 +1004,11 @@ struct GraphViewer {
   GraphViewer(const GraphViewer&) = delete;
   void operator=(const GraphViewer&) = delete;
 };
+
+struct Path {
+  PathString ToPathString() const noexcept { return g_host->Path__ToPathString(this); }
+};
```
Comment (RyanUnderhill) [inline onnxruntime/core/providers/shared_library/provider_interfaces.h:1010]: Needs a 'PROVIDER_DISALLOW_ALL(Path)' to catch errors.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h [hunk 3]
```diff
@@ -129,7 +129,7 @@ class TensorrtExecutionProvider : public IExecutionProvider {
   OrtMutex tensorrt_mu_;
   int device_id_;
   AllocatorPtr allocator_;
-  mutable int subgraph_id_ = 0;
+  mutable char model_path_[4096];
```
Comment (RyanUnderhill) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h:132]: Where does the 4096 come from? (put a comment explaining it if it's arbitrary?)

Comment (stevenlix) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.h:132]: That's the maximum possible path length. TRT parser is not willing to use wchar_t in Windows due to the code change effort, so we convert wchar_t to char at ORT side before send the path to the parser.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**